### PR TITLE
feat: Enable codec switching

### PR DESCRIPTION
**Pull request by dave-nicholas**
Closes: https://github.com/shaka-project/shaka-player/issues/1528
Closes: https://github.com/shaka-project/shaka-player/issues/1567
Closes: https://github.com/shaka-project/shaka-player/issues/4379
Closes: https://github.com/shaka-project/shaka-player/issues/5306

### CODE REVIEW

**Comment by avelad** [comment]
I am on paternity leave, but I want to review this when I return (1 week approx).

File: lib/util/stream_utils.js [hunk 1]
```diff
@@ -33,10 +34,16 @@ shaka.util.StreamUtils = class {
...
    * @param {number} preferredAudioChannelCount
    * @param {!Array.<string>} preferredDecodingAttributes
+   * @param {shaka.config.CodecSwitchingStrategy} codecSwitchingStrategy
    */
   static chooseCodecsAndFilterManifest(manifest, preferredVideoCodecs,
       preferredAudioCodecs, preferredAudioChannelCount,
-      preferredDecodingAttributes) {
+      preferredDecodingAttributes, codecSwitchingStrategy) {
+    // If enforcing a codec switch strategy skip filtration
+    if (codecSwitchingStrategy !==
```
[the remark this answers is not in the document]
Comment (dave-nicholas) [reply lib/util/stream_utils.js:43]: This PR is just providing the ability to enable codec switching. I am currently working on the second part of this which is the logic which is implementing the logic that @sky-hugolima is describing. https://github.com/shaka-project/shaka-player/pull/4766#issuecomment-1589101381

[the remark this answers is not in the document]
Comment (avelad) [reply lib/util/stream_utils.js:43]: OK, so I hope you will upload the changes to this PR. (I'd prefer all of this changes in the same PR)

[the remark this answers is not in the document]
Comment (dave-nicholas) [reply lib/util/stream_utils.js:48]: I have added the codec selection logic, if this is what you had in mind I will perform more testing and I will write the units for those methods.

[the remark this answers is not in the document]
Comment (avelad) [reply lib/util/stream_utils.js:48]: The truth is that it does not fit with what I had in mind. We want the filtering to be done by shaka itself, by default (if someone wants to customize it, it can be made customizable). So the DISABLED option should not exist. The default option should be SMOOTH except on devices we know where it doesn't work that default should be RELOAD. In your commit the option of being able to perform external filtering is added, but there is no logic in Shaka.

[the remark this answers is not in the document]
Comment (dave-nicholas) [reply lib/util/stream_utils.js:48]: The filtering logic lives inside shaka within the methods by default `filterAudioTracksForCodecSwitching_` and `filterVideoTracksForCodecSwitching_` as per https://github.com/shaka-project/shaka-player/pull/4766#discussion_r1042049562, the comparator callback is for customisation only (unless I am not understanding correctly what you are saying). 

With regards to being enabled by default, some LG devices that I have tested on _support_ `changeType` but don't seem to handle it gracefully. Should we maintain a matrix of known supported device/OS cut offs?

[the remark this answers is not in the document]
Comment (avelad) [reply lib/util/stream_utils.js:48]: Currently we already have things in the configuration that change depending on the device/browser, for example https://github.com/shaka-project/shaka-player/blob/main/lib/util/player_configuration.js#L222 So I think in this case you can do the same. If in the case of WebOS changeType does not work we must put RELOAD by default.

File: lib/util/stream_utils.js [hunk 2]
```diff
@@ -76,6 +102,119 @@ shaka.util.StreamUtils = class {
...
+    for (const variant of variants) {
+      const lang = variant.audio.language;
+      const bandwidth = variant.audio.bandwidth || 0;
+      const channelCount = variant.audio.channelsCount;
+      const codecs = variant.audio.codecs;
+      const width = variant.video.width;
+      const videoCodecs = variant.video.codecs;
+      // add codecs and width to key
+      const trackKey =
+        `${lang}-${channelCount}-${width}-${videoCodecs}`;
```
Comment (avelad) [inline lib/util/stream_utils.js:135]: Why are video parameters used to get the audio key?

File: lib/util/stream_utils.js [hunk 3]
```diff
@@ -76,6 +102,119 @@ shaka.util.StreamUtils = class {
...
+    const variantMap = new Map();
+
+    for (const variant of variants) {
+      const audiolang = variant.audio.language;
+      const videoBandwidth = variant.video.bandwidth || 0;
+      const channelCount = variant.audio.channelsCount;
+      const codecs = variant.audio.codecs;
+      const width = variant.video.width;
+      const trackKey =
+        `${audiolang}-${channelCount}-${width}`;
```
Comment (avelad) [inline lib/util/stream_utils.js:190]: Why are audio parameters used to get the video key?

File: lib/util/stream_utils.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (avelad) [reply lib/util/stream_utils.js:48]: @dave-nicholas I have shared to you on slack (video-dev) some changes about the idea of codec swithcing, I hope they help

File: karma.conf.js [hunk 4]
```diff
@@ -277,7 +277,7 @@ module.exports = (config) => {
       // stable.
       clientDisplayNone: true,
       // Only capture the client's logs if the settings want logging.
-      captureConsole: !!settings.logging && settings.logging != 'none',
+      captureConsole: true, // !!settings.logging && settings.logging != 'none',
```
Comment (avelad) [inline karma.conf.js:280]: Revert it!

File: externs/shaka/player.js [hunk 5]
```diff
@@ -1225,14 +1225,18 @@ shaka.extern.StreamingConfiguration;
...
  * @typedef {{
+ *   codecSwitchingStrategy: shaka.config.CodecSwitchingStrategy,
  *   sourceBufferExtraFeatures: string,
  *   forceTransmux: boolean,
 *    insertFakeEncryptionInInit: boolean
  * }}
  *
  * @description
  *   Media source configuration.
- *
```
Comment (avelad) [inline externs/shaka/player.js:1235]: Keep this white line, please

File: demo/common/message_ids.js [hunk 6]
```diff
@@ -0,0 +1,312 @@
+/*! @license
```
Comment (avelad) [inline demo/common/message_ids.js:1]: Please remove this file!

File: demo/locales/source.json [hunk 7]
```diff
@@ -0,0 +1,1154 @@
+{
```
Comment (avelad) [inline demo/locales/source.json:1]: Please remove this file!

File: lib/util/player_configuration.js [hunk 8]
```diff
@@ -309,11 +310,26 @@ shaka.util.PlayerConfiguration = class {
...
+    // Some devices are known not to support `MediaSource.changeType`
+    // well. These devices should use the reload strategy. If a device
+    // reports that it supports `changeType` but support it reliabley
+    // it should be added to this list.
+    if (shaka.util.Platform.isTizen2() ||
+        shaka.util.Platform.isTizen3() ||
+        shaka.util.Platform.isTizen4() ||
+        shaka.util.Platform.isWebOS3() ||
+        shaka.util.Platform.isWebOS4() ||
+        shaka.util.Platform.isWebOS5()) {
```
Comment (avelad) [inline lib/util/player_configuration.js:328]: Move it to shaka.util.Platform, similar to `supportsSequenceMode`

File: test/test/codec_switching/codec_switching_integration.js [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+    eventManager.release();
+    await player.destroy();
+  });
+
+  afterAll(() => {
+    document.body.removeChild(video);
+  });
+
+  describe('for audio', () => {
+    it('can switch codecs', async () => {
```
Comment (avelad) [inline test/test/codec_switching/codec_switching_integration.js:49]: Add another test for SMOOTH (only on supported devices)

File: lib/util/platform.js [hunk 10]
```diff
@@ -480,6 +481,27 @@ shaka.util.Platform = class {
...
 
+  /**
+   * Returns codec switching strategy by known relaible device support.
+   *
+   * Some devices are known not to support `MediaSource.changeType`
+   * well. These devices should use the reload strategy. If a device
+   * reports that it supports `changeType` but support it reliabley
+   * it should be added to this list.
+   *
+   * @return {shaka.config.CodecSwitchingStrategy}
```
Comment (avelad) [inline lib/util/platform.js:492]: returns a boolean please

Comment (avelad) [reply lib/util/platform.js]: Done!

File: lib/media/media_source_engine.js [hunk 11]
```diff
@@ -1712,6 +1740,13 @@ shaka.media.MediaSourceEngine = class {
       await sourceBufferAdded;
     } finally {
       this.reloadingMediaSource_ = false;
+
+      this.eventManager_.listenOnce(this.video_, 'canplay', () => {
```
Comment (theodab) [inline lib/media/media_source_engine.js:1744]: I think it's possible that `this.eventManager_` and `this.video_` could be null by the time the method gets here (or by the time the event manager is called), if the user loads a new asset at just the wrong time.

So this might want a call to `this.destroyer_.ensureNotDestroyed()`, before the event manager is called and inside the callback.

File: lib/media/media_source_engine.js [hunk 12]
```diff
@@ -1673,7 +1702,6 @@ shaka.media.MediaSourceEngine = class {
       }
       await Promise.all(cleanup);
       this.transmuxers_ = {};
-      this.queues_ = {};
```
Comment (theodab) [inline lib/media/media_source_engine.js:1676]: Why was this removed? I assume so that operations can continue being added, but what about the case where we are still set to RESET mode?

File: lib/media/media_source_engine.js [hunk 13]
```diff
@@ -1789,6 +1826,81 @@ shaka.media.MediaSourceEngine = class {
...
 
+  /**
+   * Returns true if it's necessary codec switch to load the new stream.
+   *
+   * @param {shaka.util.ManifestParserUtils.ContentType} contentType
+   * @param {shaka.extern.Stream} stream
+   * @return {boolean}
+   * @private
+   */
+  isNecessaryCodecSwitch_(contentType, stream) {
```
Comment (theodab) [inline lib/media/media_source_engine.js:1837]: Nit: `isNecessaryCodecSwitch_` -> `isCodecSwitchNecessary_`

File: lib/media/media_source_engine.js [hunk 14]
```diff
@@ -1789,6 +1826,81 @@ shaka.media.MediaSourceEngine = class {
...
+    const MimeUtils = shaka.util.MimeUtils;
+    const currentCodec = MimeUtils.getCodecBase(
+        MimeUtils.getCodecs(this.sourceBufferTypes_[contentType]));
+    const currentBasicType = MimeUtils.getBasicType(
+        this.sourceBufferTypes_[contentType]);
+
+    /** @type {?shaka.extern.Transmuxer} */
+    let transmuxer;
```
Comment (theodab) [inline lib/media/media_source_engine.js:1845]: Is there any reason you are defining the local variable here, when it is only used inside the 
`if (transmuxerPlugin) {}` clause below?

File: lib/media/media_source_engine.js [hunk 15]
```diff
@@ -1789,6 +1826,81 @@ shaka.media.MediaSourceEngine = class {
...
+            stream.mimeType, stream.codecs);
+    const TransmuxerEngine = shaka.transmuxer.TransmuxerEngine;
+    if (needTransmux) {
+      const transmuxerPlugin =
+          TransmuxerEngine.findTransmuxer(newMimeTypeWithAllCodecs);
+      if (transmuxerPlugin) {
+        transmuxer = transmuxerPlugin();
+        newMimeType =
+            transmuxer.convertCodecs(contentType, newMimeTypeWithAllCodecs);
+        if (transmuxer) {
```
Comment (theodab) [inline lib/media/media_source_engine.js:1865]: Similarly, once we get here, `transmuxer` is always going to be non-null, right?

File: lib/media/media_source_engine.js [hunk 16]
```diff
@@ -1789,6 +1826,81 @@ shaka.media.MediaSourceEngine = class {
...
+        }
+      }
+    }
+
+    const newCodec = MimeUtils.getCodecBase(
+        MimeUtils.getCodecs(newMimeType));
+    const newBasicType = MimeUtils.getBasicType(newMimeType);
+
+    // Current/new codecs base and basic type match then no need to switch
+    if (currentCodec === newCodec && currentBasicType === newBasicType) {
```
Comment (theodab) [inline lib/media/media_source_engine.js:1876]: Nit: `return currentCodec !== newCodec || currentBasicType !== newBasicType;`

File: lib/media/media_source_engine.js [hunk 17]
```diff
@@ -1789,6 +1826,81 @@ shaka.media.MediaSourceEngine = class {
...
+
+  /**
+   * Returns true if it's necessary reset the media source to load the
+   * new stream.
+   *
+   * @param {shaka.util.ManifestParserUtils.ContentType} contentType
+   * @param {shaka.extern.Stream} stream
+   * @return {boolean}
+   */
+  isNecessaryResetMediaSource(contentType, stream) {
```
Comment (theodab) [inline lib/media/media_source_engine.js:1890]: Nit: `isNecessaryResetMediaSource` -> `isResetMediaSourceNecessary`

File: lib/media/media_source_engine.js [hunk 18]
```diff
@@ -1789,6 +1826,81 @@ shaka.media.MediaSourceEngine = class {
...
+    if (!this.isNecessaryCodecSwitch_(contentType, stream)) {
+      return false;
+    }
+
+    if (this.config_.codecSwitchingStrategy ===
```
Comment (theodab) [inline lib/media/media_source_engine.js:1895]: Nit:

```js
return this.config_.codecSwitchingStrategy !== shaka.config.CodecSwitchingStrategy.SMOOTH ||
  !shaka.media.Capabilities.isChangeTypeSupported() ||
  this.needSplitMuxedContent_;
```

File: lib/player.js [hunk 19]
```diff
@@ -2277,8 +2279,15 @@ shaka.Player = class extends shaka.util.FakeEventTarget {
       this.onAbrStatusChanged_();
     }
 
-    // Re-filter the manifest after streams have been chosen.
-    this.filterManifestByCurrentVariant_();
+    // // Re-filter the manifest after streams have been chosen.
+    // // Note: Skip if an audio codec switching strategy is being enforced.
+    // if (this.config_.mediaSource.codecSwitchingStrategy ===
```
Comment (theodab) [inline lib/player.js:2284]: Is this meant to be part of a follow-up commit or something? Just having some random commented-out code lying around is weird.

File: externs/shaka/player.js [hunk 20]
```diff
@@ -1238,6 +1239,10 @@ shaka.extern.StreamingConfiguration;
  * @description
  *   Media source configuration.
  *
+ * @property {shaka.config.CodecSwitchingStrategy} codecSwitchingStrategy
+ *   Allow codec switching strategy. Default behavior is disabled.
```
Comment (theodab) [inline externs/shaka/player.js:1243]: The default behavior is not disabled. You commented out the code that handled that, and do not have disabled as an option on the enum. It defaults to SMOOTH if smooth codec switching is supported, RELOAD overwise.

File: lib/util/stream_utils.js [hunk 21]
```diff
@@ -38,103 +39,148 @@ shaka.util.StreamUtils = class {
...
+      }
+      const bestVariants = [];
+      variantsByResolutionMap.forEach((width, variantsByResolution) => {
+        let highestMatch = 0;
+        let matchingVariants = [];
+        for (const variant of variantsByResolution) {
+          const matchCount = preferredDecodingAttributes.filter(
+              (attribute) => variant.decodingInfos[0][attribute],
+          ).length;
+          if (matchCount>highestMatch) {
```
Comment (theodab) [inline lib/util/stream_utils.js:67]: Nit: This if (and the next) should have whitespace around the operators. E.g. `if (matchCount > highestMatch) {` and `} else if (matchCount == highestMatch) {`

File: lib/util/stream_utils.js [hunk 22]
```diff
@@ -404,12 +321,20 @@ shaka.util.StreamUtils = class {
...
+   * @param {shaka.config.CodecSwitchingStrategy=} codecSwitchingStrategy
    */
-  static async filterManifest(drmEngine, currentVariant, manifest) {
+  static async filterManifest(drmEngine, currentVariant, manifest,
+      codecSwitchingStrategy=shaka.config.CodecSwitchingStrategy.RELOAD) {
     await shaka.util.StreamUtils.filterManifestByMediaCapabilities(manifest,
         manifest.offlineSessionIds.length > 0);
-    shaka.util.StreamUtils.filterManifestByCurrentVariant(
-        currentVariant, manifest);
+    // const Strategy = shaka.config.CodecSwitchingStrategy;
```
Comment (theodab) [inline lib/util/stream_utils.js:330]: More commented out code?

**Review by theodab — CHANGES_REQUESTED** [review]

File: lib/util/stream_utils.js (hunk 22 above)
Comment (avelad) [reply lib/util/stream_utils.js]: Removed!

File: lib/player.js (hunk 19 above)
Comment (avelad) [reply lib/player.js]: Removed!

File: lib/media/media_source_engine.js (hunk 18 above)
Comment (avelad) [reply lib/media/media_source_engine.js]: Changed!

File: lib/media/media_source_engine.js (hunk 17 above)
Comment (avelad) [reply lib/media/media_source_engine.js]: Changed!

File: lib/media/media_source_engine.js (hunk 16 above)
Comment (avelad) [reply lib/media/media_source_engine.js]: Changed!

File: lib/media/media_source_engine.js (hunk 15 above)
Comment (avelad) [reply lib/media/media_source_engine.js]: Changed!

File: lib/media/media_source_engine.js (hunk 14 above)
Comment (avelad) [reply lib/media/media_source_engine.js]: Changed!

File: lib/media/media_source_engine.js (hunk 13 above)
Comment (avelad) [reply lib/media/media_source_engine.js]: Changed!

File: lib/media/media_source_engine.js (hunk 11 above)
Comment (avelad) [reply lib/media/media_source_engine.js:1746]: Added `this.destroyer_.ensureNotDestroyed()`

File: externs/shaka/player.js (hunk 20 above)
Comment (avelad) [reply externs/shaka/player.js]: Changed!

File: lib/util/stream_utils.js (hunk 21 above)
Comment (avelad) [reply lib/util/stream_utils.js]: Done!

File: lib/media/media_source_engine.js (hunk 12 above)
Comment (avelad) [reply lib/media/media_source_engine.js:1676]: In the event that we are configured in reset mode, the operations are overwritten without causing any side effects.

**Review by theodab — APPROVED** [review]
